swarm/network/simulations/discovery/dynamic_test.go

```diff
@@ -0,0 +1,538 @@
+		Action:  action,
+		Trigger: trigger,
+		Expect: &simulations.Expectation{
+			Nodes: ids[1:],
+			Check: check,
+		},
+	})
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+	close(*quitC)
+	*quitC = nil
+	sub.Unsubscribe()
+	sub = nil
+
+	// check health of all nodes
+	ctrl := newNodeCtrl(net, ids)
+	for _, n := range net.GetNodes() {
+		err := ctrl.checkHealth(n, 0)
+		if err != nil {
+			t.Fatalf("Node %v failed health check", n)
+		}
+		log.Info("node health ok", "node", n)
+	}
+
+	// asynchronously restart random nodes
+	wg := sync.WaitGroup{}
+	for i := 0; i < int(numUpDowns); i++ {
+		wg.Add(1)
+		dur := ctrl.randomDelay(0)
+		time.Sleep(dur)
+		go func() {
+			if err := ctrl.nodeUpDown(); err != nil {
+				wg.Done()
+				t.Fatal(err)
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	log.Warn("exiting test")
+}
+
+// upNodes and upAddrs: node arrays to calculate peerpot from. After restart, a node is added to the array immediately after starting and connect to bootnode
+// readyNodes: node array to choose next node to restart from. After restart, a node is added to the array after it is healthy
+type nodeCtrl struct {
+	seq                int
+	net                *simulations.Network
+	readyNodes         []discover.NodeID
+	upNodes            []discover.NodeID
+	upAddrs            [][]byte
+	addrIdx            map[discover.NodeID]string
+	pot                map[string]*network.PeerPot
+	mu                 sync.Mutex
+	healthCheckRetries int
+	healthCheckDelay   time.Duration
+}
+
+func newNodeCtrl(net *simulations.Network, nodes []discover.NodeID) *nodeCtrl {
+	ctrl := &nodeCtrl{
+		net:                net,
+		healthCheckRetries: defaultHealthCheckRetires,
+		healthCheckDelay:   defaultHealthCheckDelay,
+		addrIdx:            make(map[discover.NodeID]string),
+	}
+	for i := 0; i < len(nodes); i++ {
+		ctrl.upNodes = append(ctrl.upNodes, nodes[i])
+		ctrl.readyNodes = append(ctrl.readyNodes, nodes[i])
+		addr := network.ToOverlayAddr(ids[i].Bytes())
+		ctrl.upAddrs = append(ctrl.upAddrs, addr)
+		ctrl.addrIdx[nodes[i]] = common.Bytes2Hex(addr)
+		log.Debug("init nodeCtrl", "idx", i, "upnode", ctrl.upNodes[i], "upaddr", fmt.Sprintf("%x", ctrl.upAddrs[i]))
+	}
+	ctrl.pot = network.NewPeerPotMap(testMinProxBinSize, ctrl.upAddrs)
+	for k, p := range ctrl.pot {
+		for i, nn := range p.NNSet {
+			log.Debug("init nodeCtrl nnset", "i", i, "node", k, "nn", fmt.Sprintf("%x", nn))
+		}
+	}
+	return ctrl
+}
+
+func (self *nodeCtrl) nodeUpDown() error {
+
+	self.mu.Lock()
+
+	if len(self.readyNodes) == 1 {
+		self.mu.Unlock()
+		return errors.New("uh-oh, spaghettios: ran out of readyNodes")
+	}
+
+	// used for logging
+	self.seq++
+	seq := self.seq
+
+	// choose a random node to restart from nodes not currently stopped or getting health checked
+	nodeIdx := rand.Intn(len(self.readyNodes) - 1)
+	nodeId := self.readyNodes[nodeIdx]
+	self.readyNodes[nodeIdx] = self.readyNodes[len(self.readyNodes)-1]
```

Comment:
what's this line doing? replace the chosen random node with the last one? why?

Reply:
A "ready" node is selected from a random index of the `readyNodes` array. Overwriting that index prevents it from chosen the next round (before it's ready again).